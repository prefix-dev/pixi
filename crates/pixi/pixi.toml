[package]
name = "pixi"

[package.build.backend]
name = "pixi-build-rust"
version = "*"
