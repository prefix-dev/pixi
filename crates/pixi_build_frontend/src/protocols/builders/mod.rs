pub mod conda_build;
pub mod pixi;
pub mod rattler_build;

pub use conda_build as conda_protocol;
pub use pixi as pixi_protocol;
pub use rattler_build as rattler_build_protocol;
