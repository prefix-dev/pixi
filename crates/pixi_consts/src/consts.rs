use console::Style;
use lazy_static::lazy_static;
use std::fmt::{Display, Formatter};
use url::Url;

pub const DEFAULT_ENVIRONMENT_NAME: &str = "default";
pub const DEFAULT_FEATURE_NAME: &str = DEFAULT_ENVIRONMENT_NAME;
pub const PYPROJECT_PIXI_PREFIX: &str = "tool.pixi";

pub const PROJECT_MANIFEST: &str = "pixi.toml";
pub const PYPROJECT_MANIFEST: &str = "pyproject.toml";
pub const PROJECT_LOCK_FILE: &str = "pixi.lock";
pub const CONFIG_FILE: &str = "config.toml";
pub const PIXI_DIR: &str = ".pixi";
pub const PIXI_VERSION: &str = "0.29.0";
pub const PREFIX_FILE_NAME: &str = "pixi_env_prefix";
pub const ENVIRONMENTS_DIR: &str = "envs";
pub const SOLVE_GROUP_ENVIRONMENTS_DIR: &str = "solve-group-envs";
pub const PYPI_DEPENDENCIES: &str = "pypi-dependencies";
pub const DEPENDENCIES: &str = "dependencies";
pub const TASK_CACHE_DIR: &str = "task-cache-v0";
pub const PIXI_UV_INSTALLER: &str = "uv-pixi";
pub const CONDA_PACKAGE_CACHE_DIR: &str = rattler_cache::PACKAGE_CACHE_DIR;
pub const CONDA_REPODATA_CACHE_DIR: &str = rattler_cache::REPODATA_CACHE_DIR;
pub const PYPI_CACHE_DIR: &str = "uv-cache";
pub const CONDA_PYPI_MAPPING_CACHE_DIR: &str = "conda-pypi-mapping";
pub const CACHED_ENVS_DIR: &str = "cached-envs-v0";

pub const CONDA_INSTALLER: &str = "conda";

pub const ONE_TIME_MESSAGES_DIR: &str = "one-time-messages";

/// The default channels to use for a new project.
pub const DEFAULT_CHANNELS: &[&str] = &["conda-forge"];

pub const ENVIRONMENT_FILE_NAME: &str = "pixi";

lazy_static! {
    pub static ref TASK_STYLE: Style = Style::new().blue();
    pub static ref PLATFORM_STYLE: Style = Style::new().yellow();
    pub static ref ENVIRONMENT_STYLE: Style = Style::new().magenta();
    pub static ref FEATURE_STYLE: Style = Style::new().cyan();
    pub static ref SOLVE_GROUP_STYLE: Style = Style::new().cyan();
    pub static ref DEFAULT_PYPI_INDEX_URL: Url = Url::parse("https://pypi.org/simple").unwrap();
}

pub struct CondaEmoji;

impl Display for CondaEmoji {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if console::Term::stderr().features().colors_supported() {
            write!(f, "{}", console::style("C").bold().green())
        } else {
            write!(f, "(conda)")
        }
    }
}

pub struct PypiEmoji;

impl Display for PypiEmoji {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        if console::Term::stderr().features().colors_supported() {
            write!(f, "{}", console::style("P").bold().blue())
        } else {
            write!(f, "(pypi)")
        }
    }
}
