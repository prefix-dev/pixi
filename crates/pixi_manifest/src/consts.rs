use url::Url;

pub const PYPI_DEPENDENCIES: &str = "pypi-dependencies";
pub const DEFAULT_ENVIRONMENT_NAME: &str = "default";
pub const PROJECT_MANIFEST: &str = "pixi.toml";
pub const PYPROJECT_MANIFEST: &str = "pyproject.toml";
pub const DEFAULT_FEATURE_NAME: &str = DEFAULT_ENVIRONMENT_NAME;
pub const PYPROJECT_PIXI_PREFIX: &str = "tool.pixi";

lazy_static::lazy_static! {
    pub static ref DEFAULT_PYPI_INDEX_URL: Url = Url::parse("https://pypi.org/simple").unwrap();
}
