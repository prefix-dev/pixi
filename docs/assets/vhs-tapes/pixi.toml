[workspace]
channels = ["https://prefix.dev/conda-forge"]
platforms = ["osx-arm64", "linux-64", "win-64"]

[tasks]
_global-dark = "vhs demo_global_dark.tape"
_global-light = "vhs demo_global_light.tape"
_project-dark = "vhs demo_project_dark.tape"
_project-light = "vhs demo_project_light.tape"
record-all = { depends-on = [
  "_global-light",
  "_global-dark",
  "_project-light",
  "_project-dark",
] }

[dependencies]
ffmpeg = ">=7.1.1,<8"
vhs = ">=0.7.2,<0.8"
[target.unix.dependencies]
ttyd = ">=1.7.7,<2"
