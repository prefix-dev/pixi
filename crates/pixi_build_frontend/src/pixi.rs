use std::path::{Path, PathBuf};

use crate::tool::{IsolatedToolSpec, Tool, ToolSpec};
use crate::Metadata;
use pixi_consts::consts;
use pixi_manifest::Manifest;
use rattler_conda_types::{MatchSpec, ParseStrictness::Strict};

/// A protocol that uses a pixi manifest to invoke a build backend .
#[derive(Debug)]
pub(crate) struct PixiProtocol {
    _manifest: Manifest,
    backend_spec: ToolSpec,
}

impl PixiProtocol {
    /// Constructs a new instance from a manifest.
    pub fn new(manifest: Manifest) -> Self {
        // TODO: Replace this with something that we read from the manifest.
        let backend_spec =
            IsolatedToolSpec::from_specs(vec![
                MatchSpec::from_str("pixi-build-python", Strict).unwrap()
            ])
            .into();

        Self {
            _manifest: manifest,
            backend_spec,
        }
    }

    /// Discovers a pixi project in the given source directory.
    pub fn discover(source_dir: &Path) -> miette::Result<Option<Self>> {
        if let Some(manifest_path) = find_pixi_manifest(source_dir) {
            let manifest = Manifest::from_path(&manifest_path)?;
            return Ok(Some(Self::new(manifest)));
        }
        Ok(None)
    }

    /// Returns the backend spec of this protocol.
    pub fn backend_tool(&self) -> ToolSpec {
        self.backend_spec.clone()
    }

    /// Extract metadata from the recipe.
    pub fn get_metadata(&self, _backend: &Tool) -> miette::Result<Metadata> {
        todo!("extract metadata from pixi manifest")
    }
}

/// Try to find a pixi manifest in the given source directory.
fn find_pixi_manifest(source_dir: &Path) -> Option<PathBuf> {
    let pixi_manifest_path = source_dir.join(consts::PROJECT_MANIFEST);
    if pixi_manifest_path.exists() {
        return Some(pixi_manifest_path);
    }

    let pyproject_manifest_path = source_dir.join(consts::PYPROJECT_MANIFEST);
    // TODO: Really check if this is a pixi project.
    if pyproject_manifest_path.is_file() {
        return Some(pyproject_manifest_path);
    }

    None
}
