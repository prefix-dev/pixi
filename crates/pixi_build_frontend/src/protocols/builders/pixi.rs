use std::{
    fmt::{self, Display, Formatter},
    path::{Path, PathBuf},
    sync::Arc,
};

use miette::Diagnostic;
use pixi_consts::consts;
use pixi_manifest::{Manifest, PackageManifest, PrioritizedChannel, WorkspaceManifest};
use rattler_conda_types::{ChannelConfig, MatchSpec};
use serde::{de::IntoDeserializer, Deserialize};
use thiserror::Error;
use which::Error;

use crate::{
    jsonrpc::{Receiver, Sender},
    protocols::{InitializeError, JsonRPCBuildProtocol},
    tool::{IsolatedToolSpec, ToolCacheError, ToolSpec},
    BackendOverride, InProcessBackend, ToolContext,
};
// use super::{InitializeError, JsonRPCBuildProtocol};

/// A protocol that uses a pixi manifest to invoke a build backend .
#[derive(Debug)]
pub struct ProtocolBuilder {
    source_dir: PathBuf,
    manifest_path: PathBuf,
    workspace_manifest: WorkspaceManifest,
    package_manifest: PackageManifest,
    override_backend_spec: Option<ToolSpec>,
    channel_config: Option<ChannelConfig>,
    cache_dir: Option<PathBuf>,
}

#[derive(thiserror::Error, Debug, Diagnostic)]
pub enum ProtocolBuildError {
    #[error("failed to setup a build backend, the {} could not be parsed", .0.file_name().and_then(std::ffi::OsStr::to_str).unwrap_or("manifest")
    )]
    #[diagnostic(help("Ensure that the manifest at '{}' is a valid pixi project manifest", .0.display()
    ))]
    FailedToParseManifest(PathBuf, #[diagnostic_source] miette::Report),

    #[error("the {} does not describe a package", .0.file_name().and_then(std::ffi::OsStr::to_str).unwrap_or("manifest")
    )]
    #[diagnostic(help("A [package] section is missing in the manifest"))]
    NotAPackage(PathBuf),
}

#[derive(Debug, Error, Diagnostic)]
pub enum FinishError {
    #[error(transparent)]
    #[diagnostic(transparent)]
    Init(#[from] InitializeError),
    NoBuildSection(PathBuf),
    Tool(ToolCacheError),
    SpecConversionError(pixi_spec::SpecConversionError),
}

impl Display for FinishError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            FinishError::Init(init) => write!(f, "{init}"),
            FinishError::NoBuildSection(_) => write!(f, "failed to setup a build backend, the project manifest does not contain a [build-system] section"),
            FinishError::Tool(ToolCacheError::Instantiate(tool, err)) => match err {
                Error::CannotGetCurrentDirAndPathListEmpty | Error::CannotFindBinaryPath => write!(f, "failed to setup a build backend, the backend tool '{}' could not be found", tool.display()),
                Error::CannotCanonicalize => write!(f, "failed to setup a build backend, although the backend tool  '{}' can be resolved it could not be canonicalized", tool.display()),
            },
            FinishError::Tool(ToolCacheError::Install(report)) => write!(f, "failed to setup a build backend, the backend tool could not be installed: {}", report),
            FinishError::Tool(ToolCacheError::CacheDir(report)) => write!(f, "failed to setup a build backend, the cache dir could not be discovered: {}", report),
            FinishError::SpecConversionError(err) => write!(f, "failed to setup a build backend, the backend tool spec could not be converted: {}", err),
        }
    }
}

impl ProtocolBuilder {
    /// Constructs a new instance from a manifest.
    pub fn new(
        source_dir: PathBuf,
        manifest_path: PathBuf,
        workspace_manifest: WorkspaceManifest,
        package_manifest: PackageManifest,
    ) -> Self {
        Self {
            source_dir,
            manifest_path,
            workspace_manifest,
            package_manifest,
            override_backend_spec: None,
            channel_config: None,
            cache_dir: None,
        }
    }

    /// Sets an optional backend override.
    pub fn with_backend_override(self, backend_override: Option<BackendOverride>) -> Self {
        Self {
            override_backend_spec: backend_override
                .map(BackendOverride::into_spec)
                .or(self.override_backend_spec),
            ..self
        }
    }

    /// Sets the channel configuration used by this instance.
    pub fn with_channel_config(self, channel_config: ChannelConfig) -> Self {
        Self {
            channel_config: Some(channel_config),
            ..self
        }
    }

    /// Sets the cache directory the backend should use.
    pub fn with_opt_cache_dir(self, cache_dir: Option<PathBuf>) -> Self {
        Self { cache_dir, ..self }
    }

    /// Discovers a pixi project in the given source directory.
    pub fn discover(source_dir: &Path) -> Result<Option<Self>, ProtocolBuildError> {
        if let Some(manifest_path) = find_pixi_manifest(source_dir) {
            match Manifest::from_path(&manifest_path) {
                Ok(manifest) => {
                    // Make sure the manifest describes a package.
                    let Some(package_manifest) = manifest.package else {
                        return Err(ProtocolBuildError::NotAPackage(manifest_path));
                    };

                    let builder = Self::new(
                        source_dir.to_path_buf(),
                        manifest_path,
                        manifest.workspace,
                        package_manifest,
                    );
                    return Ok(Some(builder));
                }
                Err(e) => {
                    return Err(ProtocolBuildError::FailedToParseManifest(
                        manifest_path.to_path_buf(),
                        e,
                    ));
                }
            }
        }
        Ok(None)
    }

    pub async fn finish(
        self,
        tool: Arc<ToolContext>,
        build_id: usize,
    ) -> Result<JsonRPCBuildProtocol, FinishError> {
        let channel_config = self.channel_config.unwrap_or_else(|| {
            ChannelConfig::default_with_root_dir(
                self.manifest_path
                    .parent()
                    .expect("a manifest must always reside in a directory")
                    .to_path_buf(),
            )
        });

        let tool_spec = if let Some(backend_spec) = self.override_backend_spec {
            backend_spec
        } else {
            let build_system = &self.package_manifest.build_system;
            let specs = [(
                &build_system.build_backend.name,
                &build_system.build_backend.spec,
            )]
            .into_iter()
            .chain(build_system.additional_dependencies.iter())
            .map(|(name, spec)| {
                spec.clone()
                    .try_into_nameless_match_spec(&channel_config)
                    .map(|spec| MatchSpec::from_nameless(spec, Some(name.clone())))
            })
            .collect::<Result<_, _>>()
            .map_err(FinishError::SpecConversionError)?;

            // Figure out the channels to use
            let channels = build_system.channels.clone().unwrap_or_else(|| {
                PrioritizedChannel::sort_channels_by_priority(
                    self.workspace_manifest.workspace.channels.iter(),
                )
                .cloned()
                .collect()
            });

            ToolSpec::Isolated(IsolatedToolSpec {
                specs,
                command: build_system.build_backend.name.as_source().to_string(),
                channels,
            })
        };

        let tool = tool
            .instantiate(tool_spec, &channel_config)
            .await
            .map_err(FinishError::Tool)?;

        let configuration = self
            .package_manifest
            .build_system
            .build_backend
            .additional_args
            .map_or(serde_json::Value::Null, |value| {
                let deserializer = value.into_deserializer();
                serde_json::Value::deserialize(deserializer).unwrap_or(serde_json::Value::Null)
            });

        Ok(JsonRPCBuildProtocol::setup(
            self.source_dir,
            self.manifest_path,
            configuration,
            build_id,
            self.cache_dir,
            tool,
        )
        .await?)
    }

    /// Finish the construction of the protocol with the given
    /// [`InProcessBackend`] representing the build backend.
    pub async fn finish_with_ipc(
        self,
        ipc: InProcessBackend,
        build_id: usize,
    ) -> Result<JsonRPCBuildProtocol, FinishError> {
        let configuration = self
            .package_manifest
            .build_system
            .build_backend
            .additional_args
            .map_or(serde_json::Value::Null, |value| {
                let deserializer = value.into_deserializer();
                serde_json::Value::deserialize(deserializer).unwrap_or(serde_json::Value::Null)
            });

        Ok(JsonRPCBuildProtocol::setup_with_transport(
            "<IPC>".to_string(),
            self.source_dir,
            self.manifest_path,
            configuration,
            build_id,
            self.cache_dir,
            Sender::from(ipc.rpc_out),
            Receiver::from(ipc.rpc_in),
            None,
        )
        .await?)
    }
}

/// Try to find a pixi manifest in the given source directory.
fn find_pixi_manifest(source_dir: &Path) -> Option<PathBuf> {
    let pixi_manifest_path = source_dir.join(consts::PROJECT_MANIFEST);
    if pixi_manifest_path.exists() {
        return Some(pixi_manifest_path);
    }

    let pyproject_manifest_path = source_dir.join(consts::PYPROJECT_MANIFEST);
    // TODO: Really check if this is a pixi project.
    if pyproject_manifest_path.is_file() {
        return Some(pyproject_manifest_path);
    }

    None
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::ProtocolBuilder;

    #[test]
    pub fn discover_basic_pixi_manifest() {
        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/basic");
        let manifest_path = super::find_pixi_manifest(&manifest_dir)
            .unwrap_or_else(|| panic!("No manifest found at {}", manifest_dir.display()));
        ProtocolBuilder::discover(&manifest_path).unwrap();
    }
}
