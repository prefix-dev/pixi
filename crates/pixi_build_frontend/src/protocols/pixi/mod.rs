mod protocol;

use std::path::{Path, PathBuf};

use miette::IntoDiagnostic;
use pixi_consts::consts;
use pixi_manifest::Manifest;
pub use protocol::{InitializeError, Protocol};
use rattler_conda_types::ChannelConfig;

use crate::tool::{IsolatedToolSpec, Tool, ToolSpec};

/// A protocol that uses a pixi manifest to invoke a build backend .
#[derive(Debug)]
pub(crate) struct ProtocolBuilder {
    source_dir: PathBuf,
    _manifest: Manifest,
    backend_spec: ToolSpec,
    channel_config: ChannelConfig,
}

#[derive(thiserror::Error, Debug)]
pub enum ProtocolBuildError {
    #[error("No build section found")]
    NoBuildSection,
}

impl ProtocolBuilder {
    /// Constructs a new instance from a manifest.
    pub fn new(source_dir: PathBuf, manifest: Manifest) -> Result<Self, ProtocolBuildError> {
        let backend_spec = manifest
            .build_section()
            .map(IsolatedToolSpec::from_build_section)
            .ok_or(ProtocolBuildError::NoBuildSection)?;

        Ok(Self {
            source_dir,
            _manifest: manifest,
            backend_spec: backend_spec.into(),
            channel_config: ChannelConfig::default_with_root_dir(PathBuf::new()),
        })
    }

    /// Sets the channel configuration used by this instance.
    pub fn with_channel_config(self, channel_config: ChannelConfig) -> Self {
        Self {
            channel_config,
            ..self
        }
    }

    /// Discovers a pixi project in the given source directory.
    pub fn discover(source_dir: &Path) -> miette::Result<Option<Self>> {
        if let Some(manifest_path) = find_pixi_manifest(source_dir) {
            let manifest = Manifest::from_path(manifest_path)?;
            return Ok(Some(
                Self::new(source_dir.to_path_buf(), manifest).into_diagnostic()?,
            ));
        }
        Ok(None)
    }

    /// Returns the backend spec of this protocol.
    pub fn backend_tool(&self) -> ToolSpec {
        self.backend_spec.clone()
    }

    pub async fn finish(self, tool: Tool) -> Result<Protocol, InitializeError> {
        Protocol::setup(
            self.source_dir,
            self._manifest.path,
            self.channel_config,
            tool,
        )
        .await
    }
}

/// Try to find a pixi manifest in the given source directory.
fn find_pixi_manifest(source_dir: &Path) -> Option<PathBuf> {
    let pixi_manifest_path = source_dir.join(consts::PROJECT_MANIFEST);
    if pixi_manifest_path.exists() {
        return Some(pixi_manifest_path);
    }

    let pyproject_manifest_path = source_dir.join(consts::PYPROJECT_MANIFEST);
    // TODO: Really check if this is a pixi project.
    if pyproject_manifest_path.is_file() {
        return Some(pyproject_manifest_path);
    }

    None
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::ProtocolBuilder;

    #[test]
    pub fn discover_basic_pixi_manifest() {
        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/basic");
        let manifest_path = super::find_pixi_manifest(&manifest_dir)
            .unwrap_or_else(|| panic!("No manifest found at {}", manifest_dir.display()));
        ProtocolBuilder::discover(&manifest_path).unwrap();
    }
}
