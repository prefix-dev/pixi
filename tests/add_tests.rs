mod common;

use crate::common::builders::HasDependencyConfig;
use crate::common::package_database::{Package, PackageDatabase};
use crate::common::LockFileExt;
use crate::common::PixiControl;
use pixi::consts::DEFAULT_ENVIRONMENT_NAME;
use pixi::{DependencyType, HasFeatures, SpecType};
use rattler_conda_types::{PackageName, Platform};
use serial_test::serial;
use std::str::FromStr;
use tempfile::TempDir;

/// Test add functionality for different types of packages.
/// Run, dev, build
#[tokio::test]
async fn add_functionality() {
    let mut package_database = PackageDatabase::default();

    // Add a package `foo` that depends on `bar` both set to version 1.
    package_database.add_package(Package::build("rattler", "1").finish());
    package_database.add_package(Package::build("rattler", "2").finish());
    package_database.add_package(Package::build("rattler", "3").finish());

    // Write the repodata to disk
    let channel_dir = TempDir::new().unwrap();
    package_database
        .write_repodata(channel_dir.path())
        .await
        .unwrap();

    let pixi = PixiControl::new().unwrap();

    pixi.init()
        .with_local_channel(channel_dir.path())
        .await
        .unwrap();

    // Add a package
    pixi.add("rattler==1").await.unwrap();
    pixi.add("rattler==2")
        .set_type(DependencyType::CondaDependency(SpecType::Host))
        .await
        .unwrap();
    pixi.add("rattler==3")
        .set_type(DependencyType::CondaDependency(SpecType::Build))
        .await
        .unwrap();

    let lock = pixi.lock_file().await.unwrap();
    assert!(lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "rattler==3"));
    assert!(!lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "rattler==2"));
    assert!(!lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "rattler==1"));

    // remove the package, using matchspec
    pixi.remove("rattler==1").await.unwrap();
    let lock = pixi.lock_file().await.unwrap();
    assert!(!lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "rattler==1"));
}

/// Test that we get the union of all packages in the lockfile for the run, build and host
#[tokio::test]
async fn add_functionality_union() {
    let mut package_database = PackageDatabase::default();

    // Add a package `foo` that depends on `bar` both set to version 1.
    package_database.add_package(Package::build("rattler", "1").finish());
    package_database.add_package(Package::build("libcomputer", "1.2").finish());
    package_database.add_package(Package::build("libidk", "3.1").finish());

    // Write the repodata to disk
    let channel_dir = TempDir::new().unwrap();
    package_database
        .write_repodata(channel_dir.path())
        .await
        .unwrap();

    let pixi = PixiControl::new().unwrap();

    pixi.init()
        .with_local_channel(channel_dir.path())
        .await
        .unwrap();

    // Add a package
    pixi.add("rattler").await.unwrap();
    pixi.add("libcomputer")
        .set_type(DependencyType::CondaDependency(SpecType::Host))
        .await
        .unwrap();
    pixi.add("libidk")
        .set_type(DependencyType::CondaDependency(SpecType::Build))
        .await
        .unwrap();

    // Toml should contain the correct sections
    // We test if the toml file that is saved is correct
    // by checking if we get the correct values back in the manifest
    // We know this works because we test the manifest in another test
    // Where we check if the sections are put in the correct variables
    let project = pixi.project().unwrap();

    // Should contain all added dependencies
    let dependencies = project
        .default_environment()
        .dependencies(Some(SpecType::Run), Some(Platform::current()));
    let (name, _) = dependencies.into_specs().next().unwrap();
    assert_eq!(name, PackageName::try_from("rattler").unwrap());
    let host_deps = project
        .default_environment()
        .dependencies(Some(SpecType::Host), Some(Platform::current()));
    let (name, _) = host_deps.into_specs().next().unwrap();
    assert_eq!(name, PackageName::try_from("libcomputer").unwrap());
    let build_deps = project
        .default_environment()
        .dependencies(Some(SpecType::Build), Some(Platform::current()));
    let (name, _) = build_deps.into_specs().next().unwrap();
    assert_eq!(name, PackageName::try_from("libidk").unwrap());

    // Lock file should contain all packages as well
    let lock = pixi.lock_file().await.unwrap();
    assert!(lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "rattler==1"));
    assert!(lock.contains_match_spec(
        DEFAULT_ENVIRONMENT_NAME,
        Platform::current(),
        "libcomputer==1.2"
    ));
    assert!(lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "libidk==3.1"));
}

/// Test adding a package for a specific OS
#[tokio::test]
async fn add_functionality_os() {
    let mut package_database = PackageDatabase::default();

    // Add a package `foo` that depends on `bar` both set to version 1.
    package_database.add_package(
        Package::build("rattler", "1")
            .with_subdir(Platform::LinuxS390X)
            .finish(),
    );

    // Write the repodata to disk
    let channel_dir = TempDir::new().unwrap();
    package_database
        .write_repodata(channel_dir.path())
        .await
        .unwrap();

    let pixi = PixiControl::new().unwrap();

    pixi.init()
        .with_local_channel(channel_dir.path())
        .await
        .unwrap();

    // Add a package
    pixi.add("rattler==1")
        .set_platforms(&[Platform::LinuxS390X])
        .set_type(DependencyType::CondaDependency(SpecType::Host))
        .await
        .unwrap();

    let lock = pixi.lock_file().await.unwrap();
    assert!(lock.contains_match_spec(DEFAULT_ENVIRONMENT_NAME, Platform::LinuxS390X, "rattler==1"));
}

/// Test the `pixi add --pypi` functionality
#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
#[cfg_attr(not(feature = "slow_integration_tests"), ignore)]
#[serial]
async fn add_pypi_functionality() {
    let pixi = PixiControl::new().unwrap();

    pixi.init().await.unwrap();

    // Add python
    pixi.add("python")
        .set_type(DependencyType::CondaDependency(SpecType::Run))
        .with_install(false)
        .await
        .unwrap();

    // Add a pypi package but without installing should fail
    pixi.add("pipx")
        .set_type(DependencyType::PypiDependency)
        .with_install(false)
        .await
        .unwrap_err();

    // Add a pypi package
    pixi.add("pipx")
        .set_type(DependencyType::PypiDependency)
        .with_install(true)
        .await
        .unwrap();

    // Add a pypi package to a target
    pixi.add("boltons")
        .set_type(DependencyType::PypiDependency)
        .with_install(true)
        .set_platforms(&[Platform::Osx64])
        .await
        .unwrap();

    // Add a pypi package to a target
    pixi.add("pytest[all]")
        .set_type(DependencyType::PypiDependency)
        .set_platforms(&[Platform::Linux64])
        .with_install(true)
        .await
        .unwrap();

    let lock = pixi.lock_file().await.unwrap();
    assert!(lock.contains_pypi_package(DEFAULT_ENVIRONMENT_NAME, Platform::current(), "pipx"));
    assert!(lock.contains_pep508_requirement(
        DEFAULT_ENVIRONMENT_NAME,
        Platform::Osx64,
        pep508_rs::Requirement::from_str("boltons").unwrap()
    ));
    assert!(lock.contains_pep508_requirement(
        DEFAULT_ENVIRONMENT_NAME,
        Platform::Linux64,
        pep508_rs::Requirement::from_str("pytest[all]").unwrap(),
    ));

    // Add a pypi package with a git url
    pixi.add("requests @ git+https://github.com/psf/requests.git")
        .set_type(DependencyType::PypiDependency)
        .set_platforms(&[Platform::Linux64])
        .with_install(true)
        .await
        .unwrap();

    pixi.add("isort @ git+https://github.com/PyCQA/isort@c655831799765e9593989ee12faba13b6ca391a5")
        .set_type(DependencyType::PypiDependency)
        .set_platforms(&[Platform::Linux64])
        .with_install(true)
        .await
        .unwrap();

    pixi.add("pytest @ https://github.com/pytest-dev/pytest/releases/download/8.2.0/pytest-8.2.0-py3-none-any.whl")
        .set_type(DependencyType::PypiDependency)
        .set_platforms(&[Platform::Linux64])
        .with_install(true)
        .await
        .unwrap();

    let lock = pixi.lock_file().await.unwrap();
    assert!(lock.contains_pypi_package(DEFAULT_ENVIRONMENT_NAME, Platform::Linux64, "requests"));
    assert!(lock.contains_pypi_package(DEFAULT_ENVIRONMENT_NAME, Platform::Linux64, "isort"));
    assert!(lock.contains_pypi_package(DEFAULT_ENVIRONMENT_NAME, Platform::Linux64, "pytest"));
}

/// Test the sdist support for pypi packages
#[tokio::test(flavor = "multi_thread", worker_threads = 1)]
#[cfg_attr(not(feature = "slow_integration_tests"), ignore)]
#[serial]
async fn add_sdist_functionality() {
    let pixi = PixiControl::new().unwrap();

    pixi.init().await.unwrap();

    // Add python
    pixi.add("python")
        .set_type(DependencyType::CondaDependency(SpecType::Run))
        .with_install(true)
        .await
        .unwrap();

    // Add the sdist pypi package
    pixi.add("sdist")
        .set_type(DependencyType::PypiDependency)
        .with_install(true)
        .await
        .unwrap();
}
