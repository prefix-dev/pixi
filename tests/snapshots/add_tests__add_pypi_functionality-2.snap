---
source: tests/add_tests.rs
expression: pixi.project().unwrap().manifest.target
---
{
    PixiSpanned {
        span: Some(
            303..309,
        ),
        value: Platform(
            Osx64,
        ),
    }: TargetMetadata {
        dependencies: {},
        host_dependencies: None,
        build_dependencies: None,
        pypi_dependencies: Some(
            {
                PackageName {
                    source: "boto3",
                    normalized: "boto3",
                }: PyPiRequirement {
                    version: Some(
                        VersionSpecifiers(
                            [
                                VersionSpecifier {
                                    operator: GreaterThanEqual,
                                    version: Version {
                                        epoch: 0,
                                        release: [
                                            1,
                                            33,
                                        ],
                                        pre: None,
                                        post: None,
                                        dev: None,
                                        local: None,
                                    },
                                },
                            ],
                        ),
                    ),
                    extras: None,
                },
            },
        ),
        activation: None,
        tasks: {},
    },
    PixiSpanned {
        span: Some(
            355..363,
        ),
        value: Platform(
            Linux64,
        ),
    }: TargetMetadata {
        dependencies: {},
        host_dependencies: None,
        build_dependencies: None,
        pypi_dependencies: Some(
            {
                PackageName {
                    source: "pytest",
                    normalized: "pytest",
                }: PyPiRequirement {
                    version: None,
                    extras: Some(
                        [
                            "all",
                        ],
                    ),
                },
                PackageName {
                    source: "requests",
                    normalized: "requests",
                }: PyPiRequirement {
                    version: Some(
                        VersionSpecifiers(
                            [
                                VersionSpecifier {
                                    operator: GreaterThanEqual,
                                    version: Version {
                                        epoch: 0,
                                        release: [
                                            2,
                                            8,
                                            1,
                                        ],
                                        pre: None,
                                        post: None,
                                        dev: None,
                                        local: None,
                                    },
                                },
                                VersionSpecifier {
                                    operator: EqualStar,
                                    version: Version {
                                        epoch: 0,
                                        release: [
                                            2,
                                            8,
                                        ],
                                        pre: None,
                                        post: None,
                                        dev: None,
                                        local: None,
                                    },
                                },
                            ],
                        ),
                    ),
                    extras: Some(
                        [
                            "security",
                            "tests",
                        ],
                    ),
                },
            },
        ),
        activation: None,
        tasks: {},
    },
}
